### ISSUE TITLE
(inputs.gnmi): Parsing field names incorrectly

### ISSUE THREAD

**Issue by greenfox878** [influxdata/telegraf#14530]
### Relevant telegraf.conf

```toml
[[inputs.gnmi]]
  addresses = ["gnmi_dut:xxx"]

  ## define credentials
  username = "xxxxx"
  password = "xxxxx"

  encoding = "json_ietf"
  tagexclude = ["path"]


  [inputs.gnmi.tags]
    test_tag = "test"

  # Define additional aliases to map encoding paths to measurement names
  [inputs.gnmi.aliases]
    origin = "openconfig"
    ifcounters = "/interfaces/interface/state/counters"

  [[inputs.gnmi.subscription]]
    name = "ifcounters"
    origin = "openconfig"
    path = "/interfaces/interface/state/counters"
    subscription_mode = "sample"
    sample_interval = "10s"
```

### Logs from Telegraf

```text
no error messages
```

### System info

Telegraf 1.29.1, Ubuntu 22.04.3 LTS (latest updates)

### Steps to reproduce

1. Download compiled telegraf release "wget https://dl.influxdata.com/telegraf/releases/telegraf-1.29.1_linux_amd64.tar.gz", untar
2. Run with the command "./telegraf-1.29.1/usr/bin/telegraf --config ./test_telegraf_conifg.cfg --debug"

### Expected behavior

Correct output. v1.28.5 or older
```
ifcounters,host=test_telegraf_host,name=Ethernet40,source=gnmi_dut,test_tag=test in_broadcast_pkts=0i,in_discards=0i,in_errors=0i,in_fcs_errors=0i,in_multicast_pkts=0i,in_octets=0i,in_pkts=0i,in_unicast_pkts=0i,out_broadcast_pkts=0i,out_discards=0i,out_errors=0i,out_multicast_pkts=0i,out_octets=0i,out_pkts=0i,out_unicast_pkts=0i 1704442117721474264
ifcounters,host=test_telegraf_host,name=Ethernet26,source=gnmi_dut,test_tag=test in_broadcast_pkts=0i,in_discards=0i,in_errors=0i,in_fcs_errors=0i,in_multicast_pkts=0i,in_octets=0i,in_pkts=0i,in_unicast_pkts=0i,out_broadcast_pkts=0i,out_discards=0i,out_errors=0i,out_multicast_pkts=0i,out_octets=0i,out_pkts=0i,out_unicast_pkts=0i 1704442117721474264
```

### Actual behavior

Corrupted outptu. v1.29.1:
```
ifcounters,host=test_telegraf_host,name=Ethernet40,source=gnmi_dut,test_tag=test t_pkts=0i,in_errors=0i,rs=0i,in_octets=0i,in_pkts=0i,pkts=0i,st_pkts=0i,s=0i,out_errors=0i,out_octets=0i,out_pkts=0i,_pkts=0i 1704442117721474264
ifcounters,host=test_telegraf_host,name=Ethernet26,source=gnmi_dut,test_tag=test t_pkts=0i,in_errors=0i,rs=0i,in_octets=0i,in_pkts=0i,pkts=0i,st_pkts=0i,s=0i,out_errors=0i,out_octets=0i,out_pkts=0i,_pkts=0i 1704442117721474264
```
here we see "_pkts", "t_pkts", "st_pkts" instead of "out_pkts", "out_multicast_pkts"...

**Comment by powersj** [comment]
Hi @greenfox878,

In order to help resolve this can I request that you do the following:

1. Enable debug in the agent. Either via the `--debug` flag on the CLI or add `debug = True` in the `[agent]` section of your config.
2. In your gnmi input plugin config, add `trace = True`
3. Find and provide the full message that is getting truncated. Something like [this example](https://github.com/influxdata/telegraf/blob/master/plugins/inputs/gnmi/testcases/canonical_field_names/responses.json).

With that we can dig into the logic and see what is going on.

Thanks!

**Comment by srebhan** [comment]
@greenfox878 please use `dump_responses = true` instread of `trace = True`...

**Comment by greenfox878** [comment]
Hi @powersj ,

```
2024-01-06T07:35:45Z D! [inputs.gnmi] Got update_1704442117721474264: {"update":{"timestamp":"1704442117721474264","prefix":{"elem":[{"name":"interfaces"},{"name":"interface","key":{"name":"Ethernet35"}},{"name":"state"},{"name":"counters"}]},"update":[{"path":{"elem":[{"name":"in-broadcast-pkts"}]},"val":{"uintVal":"0"}},{"path":{"elem":[{"name":"in-discards"}]},"val":{"uintVal":"0"}},{"path":{"elem":[{"name":"in-errors"}]},"val":{"uintVal":"0"}},{"path":{"elem":[{"name":"in-fcs-errors"}]},"val":{"uintVal":"0"}},{"path":{"elem":[{"name":"in-multicast-pkts"}]},"val":{"uintVal":"0"}},{"path":{"elem":[{"name":"in-octets"}]},"val":{"uintVal":"0"}},{"path":{"elem":[{"name":"in-pkts"}]},"val":{"uintVal":"0"}},{"path":{"elem":[{"name":"in-unicast-pkts"}]},"val":{"uintVal":"0"}},{"path":{"elem":[{"name":"out-broadcast-pkts"}]},"val":{"uintVal":"0"}},{"path":{"elem":[{"name":"out-discards"}]},"val":{"uintVal":"0"}},{"path":{"elem":[{"name":"out-errors"}]},"val":{"uintVal":"0"}},{"path":{"elem":[{"name":"out-multicast-pkts"}]},"val":{"uintVal":"0"}},{"path":{"elem":[{"name":"out-octets"}]},"val":{"uintVal":"0"}},{"path":{"elem":[{"name":"out-pkts"}]},"val":{"uintVal":"0"}},{"path":{"elem":[{"name":"out-unicast-pkts"}]},"val":{"uintVal":"0"}}]}}
```

**Comment by powersj** [comment]
@greenfox878 thanks for that output. I am able to reproduce with our tests. Next steps are I need to understand the existing key parsing code. It appears in your case we want to take the path base of the key, but are instead, sometimes doing [some sub-string logic here](https://github.com/influxdata/telegraf/blob/master/plugins/inputs/gnmi/handler.go#L244-L251) that is clearly wrong.

**Comment by powersj** [comment]
@greenfox878, @romanelloacn,

Can both of you please try the artifacts in #14581 and validate that the field names report correctly please?

If they do not please use `dump_responses = true` to get the lines that are not.

Thanks!

**Comment by romanelloacn** [comment]
> @greenfox878, @romanelloacn,
> 
> Can both of you please try the artifacts in #14581 and validate that the field names report correctly please?
> 
> If they do not please use `dump_responses = true` to get the lines that are not.
> 
> Thanks!

Sorry but I have no idea how to install telegraf from the cloned telegraf repository of your profile

**Comment by powersj** [comment]
> Sorry but I have no idea how to install telegraf from the cloned telegraf repository of your profile

Artifacts, pre-build binaries, will be attached to that PR in 20-30mins. No need to build anything yourself.

**Comment by romanelloacn** [comment]
Sorry for the late reply... I can confirm that is successfully working now :) I'll wait the next official telegraf update!
<img width="198" alt="image" src="https://github.com/influxdata/telegraf/assets/152176564/c4d34822-849f-4320-b084-47cd6020fdb2">

**Comment by powersj** [comment]
Thank you for taking the time to try it out!

**Comment by romanelloacn** [comment]
> Thank you for taking the time to try it out!

You're welcome :) Can you please help with the other issue I described or should I open a new issue in github?

**Comment by powersj** [comment]
> Can you please help with the other issue I described or should I open a new issue in github?

You got the empty path error only with the current release right? The test artifact did not show that right?

**Comment by romanelloacn** [comment]
> > Can you please help with the other issue I described or should I open a new issue in github?
> 
> You got the empty path error only with the current release right? The test artifact did not show that right?

Correct

**Comment by powersj** [comment]
ok good :) The current code when parsing the field names does not take the origin (e.g. `openconfig` into account currently. So when it does the replace it effectively removes everything. My new PR will take that into account and prevent the incorrect substring as well.

**Comment by romanelloacn** [comment]
Ok thanks! So I'll wait even more patiently the new release :)

**Comment by powersj** [comment]
@romanelloacn, @greenfox878,

I made some additional changes to the PR. Could I request your try the artifacts one more time to confirm I have not regressed anything?

If all looks good, we can land this and release it in Monday's release.

Thank you!

**Comment by romanelloacn** [comment]
Hi, everything seems to be working fine! 
<img width="134" alt="image" src="https://github.com/influxdata/telegraf/assets/152176564/1a3e2f94-babd-48c4-bcf0-1ae4d939a624">

I have one doubt... I see that now in the name of the measurement, the path of the subfolders of the yang tree is included. This sometimes can be a problem because in influxdb UI, if you have 3 subfolders, you can't read the actual name of the measurement. A suggestion would be to not replace the "-" in the name with "_" so you can search the actual name in the top bar of influx and select the only result that comes out even if you can't read the full name.
Currently, it's not possible because if you search for "in-octets" it doesn't come out and if you search only "octets" both out and in measurements appear but you don't know wich one of the 2 is because there is the full path in front.


In short, I would have preferred the measurement's name to not have the subfolder name and only include the subfolder name when there is a conflict with another field having the same name in the parent folder.

Hope this can help :)

**Comment by powersj** [comment]
Thank you both for confirming.
